woocommerce/woocommerce-ios
# Order Creation: Add selected product variation to a new order

Closes: #5848

## Description

In the Order Creation flow, you can now select a product variation and it will be added to the new order. The selected variation will be displayed in the Products section of the New Order screen, with the parent variable product's name and the variation's details (e.g. variation image, price, SKU). From there, you can adjust the variation quantity or remove it from the order.

This includes updates to how `NewOrderItem` stores items for the new order. Previously, we stored the selected `Product` in `NewOrderItem`; now, we only store the properties needed for displaying/creating the order item in the new order. The immediate benefit from this change is the ability to create a new order item from either a product or a product variation. However, this will also allow us to update the new order item properties with `OrderItem` objects we receive from the API when we add support for saving auto-draft/pending orders.

## Changes

The most significant changes are to `NewOrderViewModel` and `NewOrder`:

* `NewOrderItem` now stores the individual properties needed to create an `OrderItem` to send to remote, to calculate the Products total in the new order payment section, and to fetch the product or product variation details for display in each `ProductRow`.
* Because `NewOrderItem` does not contain the corresponding `Product` or `ProductVariation`, we now fetch a list of products and a list of product variations from storage. (We need them to get all the data displayed in the Products section.) Those lists are updated when a new item is added to the order, in case the stored products/variations were updated when synced from remote.
* The product/variation lists are used in `createProductRowViewModel` to create the view model for each `ProductRow` corresponding to the items in the new order.
* `addProductVariationToOrder` handles creating a `NewOrderItem` for a selected product variation and adding it to the items in the order, similar to the existing `addProductToOrder`.
* In `NewOrder`, the view model for each `ProductRow` displayed in the `ProductInOrder` view (when a product/variation is selected from the list of items in the order) is created with `NewOrderViewModel.createProductRowViewModel`.

`ProductRow` changes:

* `ProductRowViewModel` now takes a `name` property when initialized from a product variation. This allows us to show a name created from the variation's attributes in the `AddProductVariationToOrder` view (e.g. "Blue - Any Size") but show the parent product's name in the list of products in the `NewOrder` view.

`AddProductToOrder` and `AddProductVariationToOrder` changes:

* `AddProductToOrderViewModel` takes the closure `onVariationSelected` to pass to `AddProductVariationToOrderViewModel`.
* `AddProductVariationToOrderViewModel`:
   * Takes the closure `onVariationSelected`, which is called when a product variation is selected from the list.
   * Uses `createVariationName` to create a name from a variations attributes (e.g. "Blue - Any Size").
* Removes the `Product+OrderItem` extension that was previously used to create an `OrderItem` from a `Product`. This is now handled within `NewOrderItem` directly.

Networking layer change:

* `OrdersRemote` now passes the `variationID` if the order item is a product variation or the `productID` if the order item is a non-variable product. This ensures that the product variation (rather than the parent variable product) is added when the order is created remotely.

## Testing

1. Build and run the app in debug mode (to ensure the feature flag is enabled).
2. Make sure Order Creation is enabled under Settings > Experimental Features.
3. Go to the Orders tab, tap the + button, and select "Create order."
4. On the New Order screen, tap the "Add product" button.
5. On the Add Product screen, select a non-variable product and confirm it is added to the order (same behavior as before).
6. Tap the "Add product" button again, select a variable product, and then select a product variation.
7. On the New Order screen, confirm you see the selected product variation in the Products section and can adjust its quantity or remove it (like with a non-variable product added to the order).
8. When you're done, tap "Create" and confirm your selected variation appears as expected (correct variation, correct quantity) in the created order.

## Screenshots

New Order with a non-variable product (Beanie) and a product variation (Blue Hoodie):

<img src="https://user-images.githubusercontent.com/8658164/151422693-76788ac7-c9d2-448e-a226-38fbabbdbea3.png" width="300px">

https://user-images.githubusercontent.com/8658164/151420660-7c805aad-d39f-4b16-88c5-c06d95ec4f90.mp4




## Submitter Checklist

Update release notes:
